src/test/java/com/amazon/ion/CloneTest.java

```diff
@@ -266,30 +269,151 @@ public void modifyAfterCloneDoesNotChangeOriginal() {
         assertNotEquals(original, clone);
     }
 
-    @Test
-    public void readOnlyIonStructMultithreadedTest() {
+    /**
+     * Verifies that a read-only struct without nested values can be accessed via multiple threads concurrently.
+     * @param cloneStruct true if the struct should be cloned before it is made read-only.
+     */
+    public void testReadOnlyIonStructMultithreadedAccess(boolean cloneStruct) {
         // See: https://github.com/amazon-ion/ion-java/issues/629
         String ionStr = "{a:1,b:2,c:3,d:4,e:5,f:6}";
 
         IonStruct ionValue = (IonStruct)system().singleValue(ionStr);
         ionValue.makeReadOnly();
 
-        for (int i=0; i<100; i++) {
-            IonStruct clone = ionValue.clone();
-            clone.makeReadOnly();
+        for (int i = 0; i < 100; i++) {
+            IonStruct struct;
+            if (cloneStruct) {
+                struct = ionValue.clone();
+            } else {
+                struct = ionValue;
+            }
+            struct.makeReadOnly();
 
             List<CompletableFuture<Void>> waiting = new ArrayList<>();
             for (int j = 0; j < 4; j++) {
                 waiting.add(CompletableFuture.runAsync(() -> {
                     for (int k = 0; k <= 100; k++) {
-                        assertNotNull(clone.get("a"));
+                        assertNotNull(struct.get("a"));
                     }
                 }));
             }
             waiting.forEach(CompletableFuture::join);
         }
     }
 
+    @RepeatedTest(100)
+    public void readOnlyIonStructMultithreadedAccessSucceeds() {
+        testReadOnlyIonStructMultithreadedAccess(false);
+    }
```

Comment:
The case where clone occurs is more interesting for this particular bug, but not necessarily for all possible bugs. That's why I wanted to test them separately. It would be possible to introduce a bug that would affect only the non-cloned case; if that happens, I think it would be nice to have a test that fails separately.

Reply:
I can get behind having a test that fails separately. In that case I'd rather see the behavior (`clone()`, making the parent immutable) injected in some way than boolean flags.

Reply:
I updated the tests to inject the behavior and to remove layers / quantity of repetition while still producing reliable pre-fix failures on my system.